src/main/java/com/nvidia/spark/rapids/jni/kudo/KudoSerializer.java

```diff
@@ -0,0 +1,293 @@
+ *         <br/>
+ *         <code>
+ *           int pos = col<sub>i</sub> / 8; <br/>
+ *           int bit = col<sub>i</sub> % 8; <br/>
+ *           return (hasValidityBuffer[pos] & (1 << bit)) != 0;
+ *         </code>
+ *         </td>
+ *     </tr>
+ * </table>
+ *
+ * <h2>Body</h2>
+ *
+ * The body consists of three part:
+ * <ol>
+ *     <li>Validity buffers for every column with validity in depth-first ordering of schema columns. Each buffer of
+ *     each column is 4 bytes padded.
+ *     </li>
+ *     <li>Offset buffers for every column with offsets in depth-first ordering of schema columns. Each buffer of each
+ *     column is 4 bytes padded.</li>
+ *     <li>Data buffers for every column with data in depth-first ordering of schema columns. Each buffer of each
+ *     column is 4 bytes padded.</li>
+ * </ol>
+ *
+ * <h1>Serialization</h1>
+ *
+ * The serialization process writes the header first, then writes the body. There are two optimizations when writing
+ * validity buffer and offset buffer:
+ *
+ * <ol>
+ *     <li>For validity buffer, it only copies buffers without calculating an exact validity buffer. For example, when
+ *     we want to serialize rows [3, 9) of the original table, instead of calculating the exact validity buffer, it
+ *     just copies first two bytes of the validity buffer. At read time, the deserializer will know that the true
+ *     validity buffer starts from the fourth bit, since we have recorded the row offset in the header.
+ *     </li>
+ *     <li>For offset buffer, it only copies buffers without calculating an exact offset buffer. For example, when we want
+ *  *  to serialize rows [3, 9) of the original table, instead of calculating the exact offset values by subtracting
+ *  *  first value, it just copies the offset buffer values of rows [3, 9).
+ *  *  </li>
+ *  </ol>
+ */
+public class KudoSerializer {
+
+  private static final byte[] PADDING = new byte[64];
+  private static final BufferType[] ALL_BUFFER_TYPES = new BufferType[] {BufferType.VALIDITY, BufferType.OFFSET,
+      BufferType.DATA};
+
+  static {
+    Arrays.fill(PADDING, (byte) 0);
+  }
+
+  private final int flattenedColumnCount;
+
+  public KudoSerializer(Schema schema) {
+    requireNonNull(schema, "schema is null");
+    this.flattenedColumnCount = schema.getFlattenedColumnNames().length;
+  }
+
+  /**
+   * Write partition of a table to a stream.
+   * <br/>
+   *
+   * The caller should ensure that table's schema matches the schema used to create this serializer, otherwise behavior
+   * is undefined.
+   *
+   * @param table table to write
+   * @param out output stream
+   * @param rowOffset row offset in original table
+   * @param numRows number of rows to write
+   * @return number of bytes written
+   */
+  public long writeToStream(Table table, OutputStream out, int rowOffset, int numRows) {
+    HostColumnVector[] columns = null;
+    try {
+      columns = IntStream.range(0, table.getNumberOfColumns())
```

Comment:
This looks like we are writing the whole `Table` (not sliced? but this is probably what I am missing) for every single partition of such table. Is that true? In the past we had 1 copy per column, not 200 copies per column. I am not seeing where this is getting called so that could explain it.

Reply:
ah yes, that would be great to mark as "test only". It wasn't obvious to me.